### PR TITLE
internal/resource: support S3 access point URLs

### PR DESCRIPTION
**Pull request by zeleena**
Support S3 access point URLs in ARN format as a source.
This allows valid, opaque S3 URLs such as
`s3:arn:aws:s3:us-west-2:123456789012:accesspoint/test/object`
Being able to use this format will allow S3 URLs on different
partitions and lays the foundation to potentially support
multi-region access points in the future.

Fixes https://github.com/coreos/ignition/issues/1091
Signed-off-by: Zeleena Kearney <zeleenak@lyft.com>

### CODE REVIEW

**Comment by sohankunkerkar** [comment]
Thanks for your contribution. I think you may need to add similar changes to this PR:
1. https://github.com/coreos/ignition/pull/1000/commits/dd9ca61228c637f3d6ce1459fd7d13a4a6e0d2ea (`configuration-v3_4_experimental.md`)
2. https://github.com/coreos/ignition/pull/1000/commits/abfb72808df9253ee5fef782771344ed272dcb98 (`v_3_4_experimental`)

**Review by sohankunkerkar — COMMENTED** [review]
This might need a kola test coverage, but overall changes look good to me.
@bgilbert could you take another stab at it?

**Review by sohankunkerkar — APPROVED** [review]

**Review by bgilbert — COMMENTED** [review]
Thanks for the PR!  And thanks for your patience; apologies for the slow review.

I think we should probably support this as a separate `arn` URL scheme, and be a little more rigorous about doing so, e.g. by checking the service name and by supporting non-accesspoint S3 ARNs.  Feedback welcome, though.

I agree with @sohankunkerkar that we should have a kola test for this, probably as an external test.  Feel free to add one if you'd like to dig into that infrastructure, but it probably makes more sense for us to handle that part ourselves.

**Comment by zeleena** [comment]
Thanks for the review @bgilbert . I've addressed your comments and generally supported the opaque `arn:` URL scheme.

It uses the aws sdk to more rigorously check that the ARN is for S3, and supports non-accesspoint S3 ARNs as well as accesspoint S3 ARNs. I've also added comments about the expected format of the ARN resource and official AWS documentation that supports those expectations.

Please take a look and let me know what think.

File: config/v3_4_experimental/types/url_test.go
Comment (zeleena) [reply config/v3_4_experimental/types/url_test.go]: Added `arn` as a separate URL scheme.

File: config/v3_4_experimental/types/url.go
Comment (zeleena) [reply config/v3_4_experimental/types/url.go]: Supported a new `arn` URL scheme instead.

File: config/v3_4_experimental/types/url.go
Comment (zeleena) [reply config/v3_4_experimental/types/url.go:41]: Done, added a check that the ARN service is S3 (and not some other kind) via the aws sdk.

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go]: Since I modified this to support a new `arn` URL scheme, this should now allow ARNs for an S3 object directly. I also added comments about the expected format of the ARN resource and official AWS documentation that supports those expectations.

**Comment by zeleena** [comment]
Happy new year. @bgilbert could you please take a look and let me know if this is okay to merge?

**Comment by ashleykasim** [comment]
@bgilbert could you take another look at this please?

File: internal/resource/url.go
Comment (bgilbert) [inline internal/resource/url.go:435]: Let's just return the canned `ErrInvalidS3ARN` errors rather than creating our own errors here.  Validation should have caught these before we got here anyway.

File: internal/resource/url.go
Comment (bgilbert) [inline internal/resource/url.go:424]: In the ARN case, I think the above code passes an empty bucket name to `GetBucketRegion`, which seems incorrect.

More broadly, the `arn` case currently matches the behavior of the `s3` case by assuming the same AWS partition as the one we're currently running in (or `aws` if running outside of AWS).  That seems unintuitive, since the partition is already specified in the ARN.  Let's use the partition from the ARN, provided that there's a way to do it.

**Review by bgilbert — COMMENTED** [review]
Thanks for your patience once again.  This generally looks good; just one issue around handling of AWS partitions, and a question about whether we should support version IDs with `arn`.

You'll need to rebase this to current `main`, since our CI checks have changed.  Also, in this case I think the whole PR counts as one logical change, so please squash the PR down to one commit.

I'll aim to get you a faster response in the next round.  :slightly_smiling_face:

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go:466]: I would find it unintuitive and inconsistent if I had versioning when I specify an `s3` url, but no versioning when specifying that same object as an `arn` instead.

While `arn` does have a standardized format, the [ARN docs](https://docs.aws.amazon.com/quicksight/latest/APIReference/qs-arn-format.html) suggest that the last string value, the `resource` or `resource-type` varies by service and can include a qualifier such as a version. With that caveat and mostly for the user experience reason above, I would be inclined to keep versioning available no matter how the URL is specified.

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go]: I replaced this comment with relevant bits of the comment on https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3#GetObjectInput

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go]: Done

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go]: I copied the get region logic under the url scheme cases so that the `s3` case gets the region just as before, and the `arn` case uses the partition from the ARN.

File: internal/resource/url.go
Comment (bgilbert) [inline internal/resource/url.go:442]:
```suggestion
			return configErrors.ErrInvalidS3ARN
```

File: internal/resource/url.go
Comment (bgilbert) [inline internal/resource/url.go:462]: S3 ARNs [don't include the Region field](https://docs.aws.amazon.com/AmazonS3/latest/userguide/s3-arn-format.html) so this won't be sufficient for non-`aws` partitions.  We may need to add a hardcoded map from AWS partitions (`s3arn.Partition`) to a representative region for region hinting.  (Or maybe the AWS SDK can do that for us?)

File: internal/resource/url.go
Comment (bgilbert) [reply internal/resource/url.go:468]: I read those docs as describing the general use of that field in resource-specific ARN schema.  The S3 schema in particular doesn't appear to support version IDs in that field, so AFAICT we'd be adding a custom extension.  I do see your point about supporting the full set of functionality, so on balance I'm okay with leaving this in.

The code doesn't strip off the custom query string before handing it to the ARN parser, so unlike in the `s3://` case, the query string ends up in `input.Key`.  Have you tested that fetches work with `versionId` specified?

We should also add a `TestURLValidate` test for this case.

**Review by bgilbert — COMMENTED** [review]
This looks good generally; just need to resolve issues with `versionId` and with non-`aws` partitions.

I'll be OOO for a bit, so @jlebon will handle final review and merge.  Thanks again for the PR!

**Comment by bgilbert** [comment]
Filed https://github.com/coreos/ignition/issues/1322 to track followups to this PR (kola tests and docs).

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go]: Done

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go]: AWS partitions are commonly `aws` or `aws-cn` from those docs, and there isn't a clear way to get region from those.

What I've done to address this is to use the current way of using `GetBucketRegion` for the S3 bucket and S3 bucket ARN cases since we have the bucket name available, and used the Region field for the S3 accesspoint case since [the Region field will exist for accesspoints](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-access-points.html).

**Comment by zeleena** [comment]
> I read those docs as describing the general use of that field in resource-specific ARN schema. The S3 schema in particular doesn't appear to support version IDs in that field, so AFAICT we'd be adding a custom extension. I do see your point about supporting the full set of functionality, so on balance I'm okay with leaving this in.
>
> The code doesn't strip off the custom query string before handing it to the ARN parser, so unlike in the s3:// case, the query string ends up in input.Key. Have you tested that fetches work with versionId specified?
>
> We should also add a TestURLValidate test for this case.

Not sure why I can't reply to this comment directly below it.

There is no custom query to strip off; this code uses the u.Scheme and u.Opaque to create the `fullUrl` variable, and u.Query is a separate field that contains the versionId. See how URLs are interpreted [here](https://pkg.go.dev/net/url#URL). So the query string does not end up in input.Key because it is a separate field which is not included when handed to the ARN parser.

I added test cases to cover all three cases of S3 bucket, S3 bucket ARN and S3 accesspoint ARN and their corresponding URL with versioning.

**Comment by zeleena** [comment]
Addressed the comments and squashed commits. PTAL @jlebon or @bgilbert if you return from OOO.

**Review by jlebon — APPROVED** [review]
One suggestion but LGTM as is too. Happy to leave that for a follow-up if you prefer.

**Comment by jlebon** [comment]
Also, can you confirm whether you were able to test this in AWS?

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go]: Moved to a helper function `parseARN` which contains and simplifies the logic here and included corresponding unit tests for it

**Comment by zeleena** [comment]
> Also, can you confirm whether you were able to test this?

Yes, this can still retrieve configs from S3 given a bucket in s3 or arn format as before. Multi-region access points won't be enabled until this uses aws-sdk-go-v2, which I hope to do in a follow-up PR.

Also one thing is that I had to modify my setup to include a fix for https://github.com/golang/go/issues/51706, but I imagine this was related to my local setup.

Please take a look and feel free to merge if this is good to go.

File: internal/resource/url.go
Comment (jlebon) [inline internal/resource/url.go:568]: The code below this uses indexing into this slice directly; are we 100% sure that no URL could get to here (i.e. past `arn.IsARN()` and `arn.Parse()`) such that those indexes could never cause a panic?

Would it be safer to just do sanity-checks of its length in each case (accesspoint vs not) and nicely return an error if it's not the minimum we expect?

**Review by jlebon — COMMENTED** [review]
Thanks, looks much nicer now! Have one remaining comment, otherwise this LGTM. I'll let @bgilbert also have a look.

**Comment by jlebon** [comment]
Also, it looks like there is a merge commit in the PR. Could you try rebasing your commit onto the latest main?

File: internal/resource/url.go
Comment (zeleena) [reply internal/resource/url.go]: Sure, I added some sanity length checks to this function and in the URL validation, and added test cases to check this. Unfortunately the ARN package just handles splitting the ARN parts delimited by `:`. Since resources could be from other services (ex. IAM), they don't expose a way to index or split the resource part. I added the relevant documentation https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arns-paths in a comment there as well.

**Comment by zeleena** [comment]
> Also, it looks like there is a merge commit in the PR. Could you try rebasing your commit onto the latest main?

Addressed comment and rebased. Mind taking a look now @jlebon @bgilbert ?

**Comment by zeleena** [comment]
Hi all, it's been almost two weeks since the last reply. Does this look good and is this okay to merge? @bgilbert @jlebon

File: internal/resource/url.go
Comment (bgilbert) [inline internal/resource/url.go:570]: Nit: doesn't need to be predeclared.

File: internal/resource/url_test.go
Comment (bgilbert) [inline internal/resource/url_test.go:301]:
```suggestion
			t.Errorf("#%d: expected error %+v, got %+v", i, test.out.err, err)
```

**Review by bgilbert — APPROVED** [review]
This looks great!  Thanks so much for your extensive patience in pushing this to the finish line.

There are a couple tiny nits and I'm still not convinced about the partition selection for non-accesspoint ARNs, but let's merge this now and I'll see about making a followup PR.

File: internal/resource/url.go
Comment (bgilbert) [reply internal/resource/url.go]: As is, the non-accesspoint case will fail to select the correct region if we're running in a different partition than the one specified in the ARN.  I still think a lookup table for a representative `regionHint`, indexed by partition, is the right solution here.

**Comment by bgilbert** [comment]
Nits addressed in https://github.com/coreos/ignition/pull/1355.

**Pull request merged**
